browse/services/dissemination/article_store.py

```diff
@@ -375,6 +393,58 @@ def _pdf(self, arxiv_id: Identifier, docmeta: DocMetadata, version: VersionEntry
                      [str(ps_cache_pdf), str(pdf_file)])
         return "UNAVAILABLE"
 
+    def _genpdf(self, arxiv_id: Identifier, docmeta: DocMetadata, version: VersionEntry, api: str) -> FormatHandlerReturn:
+        """Gets a PDF from the genpdf-api."""
+        if version.source_flag.cannot_pdf:
+            return "NOT_PDF"
+
+        # requests.get() cannod have timeout <= 0
+        timeout = max(1, current_app.config.get("GENPDF_API_TIMEOUT", 60))
+        url = f"{api}/pdf/{arxiv_id.ids}?timeout={timeout}"
+        logger.debug("genpdf-api request(timeout=%d): %s ", timeout, url)
+        t_start = time.perf_counter()
+        try:
+            response: requests.Response = requests.get(url, timeout=timeout, allow_redirects=False)
+        except Exception as _exc:
+            logger.warning("genpdf-api access failed", exc_info=True)
+            return "UNAVAILABLE"
+        t_end = time.perf_counter()
+        logger.debug("genpdf-api responded in %f seconds", t_end - t_start)
+
+        # Normal operation is a redirect to the bucket
+
+        match response.status_code:
+            case http.HTTPStatus.FOUND:
+                bucket_url = response.headers.get('location')
+                if not bucket_url:
+                    logger.error("Redirect did not provide location")
+                    return "UNAVAILABLE"
+                if isinstance(self.objstore, GsObjectStore):
+                    bucket_name, obj_key = from_bucket_url_to_key(bucket_url)
+                    if obj_key and bucket_name == self.objstore.bucket.name:
```

Comment:
This seems to require that the bucket sent back by genpdf-api is the same as the one configured in arxiv-browse. Is that going to always be the case? Does the flask app need a config of "GENPDF_API_BUCKET" and then have a genpdf_objstore?  

`FormhandlerReturn` can be a `FileObject` and google.cloud.stroage `Blob` is a `FileObject`

Reply:
I am not clear on this TBH. GenPDF's location is the full bucket path and that's fine. Whether or not Browse uses same bucket is probably Browse's business. Here, I basically chickened out to say that they should use the same bucket but they don't have to. 

Having separate bucket store may make sense. In some sense, it doesn't hurt to do so, so I'll do it. If we change mind, not a big deal.

> FormhandlerReturn can be a FileObject and google.cloud.stroage Blob is a FileObject

Not quite sure on this. I'll ask you in Slack.